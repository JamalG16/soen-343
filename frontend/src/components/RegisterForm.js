import React from 'react'
import { Button, Form, Grid, Header, Segment, Checkbox, Message } from 'semantic-ui-react'
import { register } from '../util/ApiUtil';
class LoginForm extends React.Component {
    state = {
        firstName: '',
        lastName: '',
        address: '',
        phone: '',
        email: '',
        password: '',
        isAdmin: false,
        submitting: false,
        success: false,
        error: false,
        errorMessage: 'The account could not be created',
    };

    handleChange = (e, { name, value }) => this.setState({ [name]: value });

    handleCheckbox = () =>  this.setState((state) => ({ isAdmin: !state.isAdmin }));

    handleSubmit = async () => {
        this.setState({ submitting: true });

        const { firstName, lastName, email, address, phone, password, isAdmin } = this.state;
        try {
            await register(firstName, lastName, email, address, phone, password, isAdmin);
            this.setState({ 
                submitting: false, 
                success: true,
            });
        } catch (err) {
            this.setState({
                submitting: false,
                error: true,
            });
        }
    };

    initializeForm = () => this.setState({
        firstName: '',
        lastName: '',
        address: '',
        phone: '',
        email: '',
        password: '',
        admin: false,
        submitting: false,
        success: false,
        errorMessage: 'The account could not be created',
    });

    render() {
        const { 
            firstName, 
            lastName, 
            address, 
            email, 
            phone, 
            password, 
            isAdmin, 
            submitting, 
            success, 
            error, 
            errorMessage,
        } = this.state;
        return (
            <div className='register-form' style={{ height: '100%' }}>
                <Grid 
                    textAlign='center' 
                    verticalAlign="middle"
                    style={{ height: '100%' }}>
                    <Grid.Column style={{ width: '80%', maxWidth: '600px' }}>
                        <Header as='h2' color='teal' textAlign='left'>
                        Register a Client
                        </Header>
                        <Form size='big' success={success} onSubmit={this.handleSubmit} error={error}>
                            <Segment stacked>
                                <Form.Input 
                                    name ="firstName" 
                                    fluid icon='user outline' 
                                    iconPosition='left' 
                                    placeholder='First Name'
                                    value={firstName}
                                    onChange={this.handleChange} 
                                    required/>
                                <Form.Input 
                                    name ="lastName" 
                                    fluid icon='user outline' 
                                    iconPosition='left' 
                                    placeholder='Last Name' 
                                    value={lastName}
                                    onChange={this.handleChange} 
                                    required/>
                                <Form.Input 
                                    name ="address" 
                                    fluid icon='home' 
                                    iconPosition='left' 
                                    placeholder='Address' 
                                    value={address}
                                    onChange={this.handleChange} 
                                    required/>
                                <Form.Input 
                                    name ="phone" 
                                    fluid icon='phone' 
                                    iconPosition='left' 
                                    placeholder='Phone Number' 
                                    value={phone}
                                    onChange={this.handleChange} 
                                    required/>
                                <Form.Input 
                                    name ="email" 
                                    fluid icon='mail' 
                                    iconPosition='left' 
                                    placeholder='E-mail address' 
                                    value={email}
                                    onChange={this.handleChange} 
                                    required/>
                                <Form.Input 
                                    name ="password"    
                                    fluid icon='lock' 
                                    iconPosition='left' 
                                    placeholder='Password' 
                                    type='password' 
                                    value={password}
                                    onChange={this.handleChange} 
                                    required/>
                                <Form.Field
                                    name="isAdmin" 
                                    control={Checkbox} 
                                    label={{ children: 'This is an administrator account' }} 
                                    checked={isAdmin}
                                    onChange={this.handleCheckbox} />
                                <Message 
                                    error  
                                    header="Registration failed" 
                                    content={errorMessage}
                                    style={{ textAlign: 'left' }} />
                                <Button 
                                    loading={submitting} 
                                    color='teal' 
                                    size='large' 
                                    type="submit"
                                    fluid>
                                        Register
                                </Button>
                            </Segment>
                            <Message 
                                success
                                header='Registration complete!' 
                                content={`An ${isAdmin ? 'administrative' : ''} account was created for ${firstName} ${lastName}`} />
                        </Form>
                        { 
                            success && 
                            <Header 
                                as='h3' 
                                className="enter-new-user"
                                onClick={this.initializeForm}>Enter another user</Header>}
                    </Grid.Column>
                </Grid>
            </div>
        )
    }
}


export default LoginForm;
