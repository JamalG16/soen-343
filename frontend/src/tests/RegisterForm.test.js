import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

import RegisterForm from '../components/RegisterForm';

Enzyme.configure({ adapter: new Adapter() });

const wrapper = shallow(<RegisterForm/>);
const form = wrapper.find('Form');
const submitHandlerSpy = jest.spyOn(wrapper.instance(), 'handleSubmit');

describe('RegisterForm', () => {
    it('renders correctly', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('does not submit if a required field is empty', () => {      
        form.simulate('submit');
        expect(submitHandlerSpy).not.toHaveBeenCalled();
    });
}); 

