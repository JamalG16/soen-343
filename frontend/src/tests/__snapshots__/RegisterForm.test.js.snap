// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RegisterForm renders correctly 1`] = `
<div
  className="login-form"
>
  <style>
    
                    body &gt; div,
                    body &gt; div &gt; div,
                    body &gt; div &gt; div &gt; div.login-form {
                        height: 100%;
                    }
                    
  </style>
  <Grid
    textAlign="center"
  >
    <GridColumn
      style={
        Object {
          "maxWidth": "600px",
          "width": "80%",
        }
      }
    >
      <Header
        as="h2"
        color="teal"
        textAlign="left"
      >
        Register a Client
      </Header>
      <Form
        as="form"
        onSubmit={[Function]}
        size="big"
        success={false}
      >
        <Segment
          stacked={true}
        >
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            icon="user outline"
            iconPosition="left"
            name="fname"
            onChange={[Function]}
            placeholder="First Name"
            required={true}
            value=""
          />
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            icon="user outline"
            iconPosition="left"
            name="lname"
            onChange={[Function]}
            placeholder="Last Name"
            required={true}
            value=""
          />
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            icon="home"
            iconPosition="left"
            name="address"
            onChange={[Function]}
            placeholder="Address"
            required={true}
            value=""
          />
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            icon="phone"
            iconPosition="left"
            name="phone"
            onChange={[Function]}
            placeholder="Phone Number"
            required={true}
            value=""
          />
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            icon="mail"
            iconPosition="left"
            name="email"
            onChange={[Function]}
            placeholder="E-mail address"
            required={true}
            value=""
          />
          <FormInput
            as={[Function]}
            control={[Function]}
            fluid={true}
            icon="lock"
            iconPosition="left"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            required={true}
            type="password"
            value=""
          />
          <FormField
            checked={false}
            control={[Function]}
            label={
              Object {
                "children": "This is an administrator account",
              }
            }
            name="admin"
            onChange={[Function]}
          />
          <Button
            as="button"
            color="teal"
            fluid={true}
            loading={false}
            role="button"
            size="large"
            type="submit"
          >
            Register
          </Button>
        </Segment>
        <Message
          content="An  account was created for  "
          header="Registration complete!"
          success={true}
        />
      </Form>
    </GridColumn>
  </Grid>
</div>
`;
